app/src/main/java/com/infomaniak/swisstransfer/upload/UploadForegroundService.kt

```diff
@@ -203,7 +206,56 @@ private fun <T> Flow<T>.rateLimit(minInterval: Duration): Flow<T> = channelFlow
         val timeSinceLastEmit = nanosSinceLastEmit.nanoseconds
         val timeToWait = minInterval - timeSinceLastEmit.coerceAtMost(minInterval)
         delay(timeToWait)
-        send(newValue) // We don't handle the case where cancellation happens after it's received.
+        // We use `sendAtomic` instead of just `send` to ensure `lastEmitElapsedNanos` is updated
+        // if the value was successfully sent while cancellation was happening.
+        // That ensures we count every value being sent, and respect the desired rate limiting.
+        // FYI, as its documentation states, `send` can deliver the value to the receiver,
+        // and throw a `CancellationException` instead of returning (if coroutine's `Job` gets cancelled).
+        sendAtomic(newValue)
         lastEmitElapsedNanos = SystemClock.elapsedRealtimeNanos()
     }
 }.buffer(Channel.RENDEZVOUS)
+
+
+/**
+ * [SendChannel.send] can throw a [CancellationException] after the value was sent,
+ * which might not be desirable if we want to take factor-in whether the value was actually
```

Comment:
You should choose between "take" and "factor-in"

Reply:
Nice catch, removed "take" that was a leftover from rewriting.